crates/store/src/state.rs

```diff
@@ -653,25 +660,72 @@ impl State {
 
     /// Lists all known nullifiers with their inclusion blocks, intended for testing.
     pub async fn list_nullifiers(&self) -> Result<Vec<(Nullifier, u32)>, DatabaseError> {
-        self.db.select_nullifiers().await
+        self.db.select_all_nullifiers().await
     }
 
     /// Lists all known accounts, with their ids, latest state hash, and block at which the account
     /// was last modified, intended for testing.
     pub async fn list_accounts(&self) -> Result<Vec<AccountInfo>, DatabaseError> {
-        self.db.select_accounts().await
+        self.db.select_all_accounts().await
     }
 
     /// Lists all known notes, intended for testing.
     pub async fn list_notes(&self) -> Result<Vec<NoteRecord>, DatabaseError> {
-        self.db.select_notes().await
+        self.db.select_all_notes().await
     }
 
-    /// Returns details for public (public) account.
+    /// Returns details for public (on-chain) account.
     pub async fn get_account_details(&self, id: AccountId) -> Result<AccountInfo, DatabaseError> {
         self.db.select_account(id).await
     }
 
+    /// Returns account states with details for public accounts.
+    pub async fn get_account_states(
+        &self,
+        account_ids: HashSet<AccountId>,
+        include_headers: bool,
+    ) -> Result<(BlockNumber, Vec<AccountStateResponse>), DatabaseError> {
+        // Lock inner state for the whole operation. We need to hold this lock to prevent the
+        // database, account tree and latest block number from changing during the operation,
+        // because changing one of them would lead to inconsistent state.
+        let inner_state = self.inner.read().await;
+
+        let infos = self.db.select_accounts_by_ids(account_ids.iter().copied().collect()).await?;
```

Comment:
We need to go to the database only if `include_headers` is true - right? Otherwise, all the info we need should be already in memory.

Reply:
Good catch, thank you!